### ISSUE TITLE
Lua type annotations

### ISSUE THREAD

**Issue by kylechui** [wez/wezterm#3132]
**Is your feature request related to a problem? Please describe.**
Seeing as Lua is a dynamically typed scripting language, it might be beneficial to include [type annotations](https://github.com/LuaLS/lua-language-server/wiki/Annotations) to provide autocomplete (reducing typos), type-checking, and potentially documentation.

**Describe the solution you'd like**
Have the keys for the configuration table be annotated, so users using [LSP](https://github.com/LuaLS/lua-language-server/) could benefit from more information.

**Describe alternatives you've considered**
The documentation on the website is very clear and helpful (in particular with having the default keys right there), but requires another window open to look back and forth between.

**Additional context**
Seeing as the codebase is written almost entirely in Rust, I'm not sure how the annotations would work (since they're typically added in Lua files). Here's some screenshots illustrating what could be possible:
![image](https://user-images.githubusercontent.com/48545987/219909122-8165664e-9243-4fcf-88d0-2ab818638006.png)
![image](https://user-images.githubusercontent.com/48545987/219909126-b44a7919-8541-41f8-9c93-0b4209013f78.png)
![image](https://user-images.githubusercontent.com/48545987/219909131-c2b869ee-783b-4fe9-a0f8-12e96dbdb389.png)

**Comment by wez** [comment]
Relevant prior discussion: https://matrix.to/#/!PirwUBcuIlTXwNveYz:matrix.org/$Yiujl4grH4m7PllU0_lDDNr4-yoHywtotC3r1drmg1E?via=matrix.org&via=kde.org&via=envs.net

**Comment by kylechui** [comment]
Hmm it says that the link was malformed... do you mind trying to ping me on matrix? (still @kylechui)

**Comment by chrisgrieser** [comment]
I also support this. There are two projects I know of that add annotations (both by folke, actually)
1. [neodev.nvim](https://github.com/folke/neodev.nvim) (Adds annotations for nvim)
2. [Emmylua.spoon](https://www.hammerspoon.org/Spoons/EmmyLua.html) (Adds annotations for Hammerspoon)

While neodev uses annotations available in nvim's code, Emmylua takes a different approach and creates dummy files with empty functions and the the respective documentation added as luadoc above, for example in the file `hs.application.lua`

```lua
local M = {}
-- ...

-- Returns the filesystem path of the app.
--
-- Parameters:
--  * None
--
-- Returns:
--  * A string containing the filesystem path of the application or nil if the path could not be determined (e.g. if the application has terminated).
---@return string
function M:path() end

-- ...
return M
```
All the annotations are saved in a folder. That folder only has to be added as a [workspace library](https://github.com/LuaLS/lua-language-server/wiki/Settings#workspacelibrary) either in the configuration for the lua lsp, or in a `.luarc.json` file (so the annotations are only available in the respective folder/workspace).

The approach of EmmyLua.spoon is quite straightforward and should be possible for WezTerm as well.

**Comment by junkblocker** [comment]
After trying to use https://github.com/nekowinston/wezterm-types ( @nekowinston ) , I realized I had to annotate the variables myself e.g.

```lua
---@type WezTermBlah
local wezterm = require "wezterm"
```

But then I found @jason0x43 's

https://github.com/jason0x43/dotfiles/tree/master/config/nvim/lua/user/types/wezterm

which did not require me to annotate the vars myself. Here are the changes that were needed for using it with nvim/lua_ls after extracting that folder at `~/dev/weztypes`

```
diff --git a/.config/nvim/lua/plugins/lsp.lua b/.config/nvim/lua/plugins/lsp.lua
index f7a0c507..f1408995 100644
--- a/.config/nvim/lua/plugins/lsp.lua
+++ b/.config/nvim/lua/plugins/lsp.lua
@@ -103,6 +103,8 @@ local modules = {
             end
 
             local runtime_path = vim.split(package.path, ";")
+            table.insert(runtime_path, "?.lua")
+            table.insert(runtime_path, "?/init.lua")
             table.insert(runtime_path, "lua/?.lua")
             table.insert(runtime_path, "lua/?/init.lua")
 
@@ -182,6 +184,8 @@ local modules = {
                 -- For others, just run the inside content of the function
                 -- after this setup
                 ["lua_ls"] = function()
+                    local libs = vim.api.nvim_get_runtime_file("", true)
+                    table.insert(libs, vim.fn.expand("~/dev/weztypes"))
                     lspconfig.lua_ls.setup(vim.tbl_deep_extend("keep", config(), {
                         root_dir = function(...)
                             -- doing this so editorconfig based formatting happens instead of stylua based
@@ -202,7 +206,7 @@ local modules = {
                                 },
                                 workspace = {
                                     checkThirdParty = false, -- prevent luassert, luafilesystem, OpenResty annoyance
-                                    library = vim.api.nvim_get_runtime_file("", true),
+                                    library = libs,
                                     maxPreload = 100000,
                                     preloadFileSize = 10000,
                                 },
```

I see a mention of @wez saying on matrix that he can generate the types automatically from the source code. If so, supposedly this is the format that's most convenient acc. to this experiment.

**Comment by chrisgrieser** [comment]
Hmm, I tried various repos I could find for `wezterm-types`, and all of them are incomplete and/or outdated. The most complete one is `justinsgithub/wezterm-types`, though it is still lacking.

Due to the size and the ongoing pace of WezTerm development, I think auto-generating it from wezterm itself would be the way to go.

**Comment by mikavilpas** [comment]
TL;DR: have been looking at the same issue the last few weeks, and generating type annotations from the docs seems to be the easiest option for me too.

I have been looking for a good way to create lua language server type annotations from rust code myself. I don't think there is a ready solution available at the moment.

I actually started writing a rust implementation for the lua_ls type annotation system (only annotation/documentation generation) and proposed it as a solution in the yazi project [here](https://github.com/sxyazi/yazi/discussions/981) and [here](https://github.com/sxyazi/yazi/pull/986), but for now it looks like the discussion has stalled.

Even if my attempt were successful, I am not convinced it would be a good solution unless there is either

1. significant buy-in from the project's developer team, or
2. significant buy-in from mlua (the rust-lua bindings crate) team, which can be tracked e.g. here https://github.com/mlua-rs/mlua/issues/392. For example, automatically type checking the annotations based on the rust code would be a great addition.

**Comment by wez** [comment]
I think there are two parts to this:

* How to conveniently annotate and expose metadata about modules and functions
* How to conveniently annotate and expose metadata about types

The former either needs some support from mlua, or needs to employ something like the Api Definition stuff in eg: [dropshot](https://docs.rs/dropshot/latest/dropshot/). dropshot is a REST API helper, rather than lua, but the concept and model could be applied to lua.  The idea is that a proc macro with declarative attributes could be used to annotate the various functions and collect them together in a supplemental metadata struct.  There could then be a `wezterm` subcommand that can emit that metadata in various forms. I think that this could be used to augment the docs as well.

The latter could leverage the existing `ConfigMeta` trait that is auto-derived for the main config struct:
https://github.com/wez/wezterm/blob/b8f94c474ce48ac195b51c1aeacf41ae049b774e/config/src/config.rs#L51-L52
The same trait could be applied to more of the various types that are part of the API.

**Comment by ZipingL** [comment]
is this being developed

**Comment by juster-0** [comment]
@wez while the feature is being developed (or maybe not, because it's too much headache), I propose meanwhile to add lua annotations "manually". Just for testing I've done a little bit and here is the result in neovim and vscode (I don't use it much and I've pressed something, so the video is not that good, but the idea is visible on the video):
https://youtu.be/26LBCVRtpco
https://youtu.be/XtX-92byKQQ

In the videos you can see that it's possible to include some documentation into the hints, and it works without problem on different LSP clients, everything that I've done to "activate" this I've added `~/.config/wezterm/.luarc.json` file with content:
```json
{
    "$schema": "https://raw.githubusercontent.com/sumneko/vscode-lua/master/setting/schema.json",
    "workspace.library": [
        "~/.local/share/wezterm/libs/lua/"
    ]
}
```
`libs/lua` just contains `.lua` files with the annotations.
I ready to spend some time to do that, if it's not just for myself.
And while automatic system is not ready it's not a big problem to update docs in the annotation files in the future.
What do you thing about this?

**Comment by digitallyserviced** [comment]
@junkblocker @junkblocker @juster-0 @wez 

This `wezterm-types` is the most up-to-date and activate repo.

https://github.com/gonstoll/wezterm-types
